sycl/include/CL/sycl/intel/sub_group.hpp

```diff
@@ -25,69 +26,167 @@ template <typename T, access::address_space Space> class multi_ptr;
+template <typename T, access::address_space Space>
+void store(multi_ptr<T, Space> dst, const T &x) {
+  using Ty = SelectBlockTy<T>;
+  using GlobalPtrTy = __attribute__((opencl_global)) Ty *;
+
+  __spirv_SubgroupBlockWriteINTEL(reinterpret_cast<GlobalPtrTy>(dst.get()),
+  // TODO: any better representation of reinterpret cast from T to Ty for x?
+                                      *((Ty *)&x));
 }
 
+template <int N, typename T, access::address_space Space>
+void store(multi_ptr<T, Space> dst, const vec<T, N> &x) {
+  using Ty = SelectBlockTy<T>;
+  using VecTy = Ty __attribute__((ext_vector_type(N)));
+  using GlobalPtrTy = __attribute__((opencl_global)) Ty *;
+
+  __spirv_SubgroupBlockWriteINTEL(reinterpret_cast<GlobalPtrTy>(dst.get()),
+  // TODO: any better representation of reinterpret cast from T to Ty for x?
+                                      *((VecTy *)&x));
+}
+
+struct GroupOpISigned {}; struct GroupOpIUnsigned {}; struct GroupOpFP {};
+
+template <class T, typename = void> struct GroupOpTag;
+
+template <class T>
+struct GroupOpTag<T, detail::enable_if_t<detail::is_sigeninteger<T>::value>> {
+  using type = GroupOpISigned;
+};
+
+template <class T>
+struct GroupOpTag<T, detail::enable_if_t<detail::is_sugeninteger<T>::value>> {
+  using type = GroupOpIUnsigned;
+};
+
+template <class T>
+struct GroupOpTag<T, detail::enable_if_t<detail::is_sgenfloat<T>::value>> {
+  using type = GroupOpFP;
+};
+
+#define CALC_BODY(Operation)                                                   \
+  using OCLTy = detail::ConvertToOpenCLType_t<T>;                              \
+  OCLTy Arg = detail::convertDataToType<T, OCLTy>(x);                          \
+  OCLTy Ret = __spirv_Group##Operation(__spv::Scope::Subgroup, O, Arg);        \
+  return detail::convertDataToType<OCLTy, T>(Ret);
+
+template <typename T, __spv::GroupOperation O>
+static T calc(GroupOpISigned, T x, intel::minimum<T> op) { CALC_BODY(SMin) }
+
+template <typename T, __spv::GroupOperation O>
+static T calc(GroupOpIUnsigned, T x, intel::minimum<T> op) { CALC_BODY(UMin) }
+
+template <typename T, __spv::GroupOperation O>
+static T calc(GroupOpFP, T x, intel::minimum<T> op) { CALC_BODY(FMin) }
+
+template <typename T, __spv::GroupOperation O>
+static T calc(GroupOpISigned, T x, intel::maximum<T> op) { CALC_BODY(SMax) }
+
+template <typename T, __spv::GroupOperation O>
+static T calc(GroupOpIUnsigned, T x, intel::maximum<T> op) { CALC_BODY(UMax) }
+
+template <typename T, __spv::GroupOperation O>
+static T calc(GroupOpFP, T x, intel::maximum<T> op) { CALC_BODY(FMax) }
+
+template <typename T, __spv::GroupOperation O>
+static T calc(GroupOpISigned, T x, intel::plus<T> op) { CALC_BODY(IAdd<T>) }
+
+template <typename T, __spv::GroupOperation O>
+static T calc(GroupOpIUnsigned, T x, intel::plus<T> op) { CALC_BODY(IAdd<T>) }
+
+template <typename T, __spv::GroupOperation O>
+static T calc(GroupOpFP, T x, intel::plus<T> op) { CALC_BODY(FAdd<T>) }
```

Comment:
One can put the whole definition into a macro

Reply:
Done